L74-java-channel-creds.md

```diff
@@ -0,0 +1,478 @@
+    .build();
+channel = ManagedChannelBuilder.forTarget("example.com").usePlaintext()
+    .build();
+channel = ManagedChannelBuilder.forAddress("example.com", 443)
+    .build();
+// the user would now:
+channel = Grpc.newChannelBuilder("example.com", new TlsChannelCredentials())
+    .build();
+channel = Grpc.newChannelBuilder("example.com", new InsecureChannelCredentials())
+    .build();
+channel = Grpc.newChannelBuilderForAddress("example.com", 443, new TlsChannelCredentials())
+    .build();
+```
+
+`newChannelBuilder()` will iterate through the `ManagedChannelProviders`
+trying to create a builder for each in turn. The first provider that succeeds
+will be used. If no provider is able to handle the credentials, then
+`newChannelBuilder()` will throw. `newChannelBuilderForAddress()` will be a
+convenience function that creates the target string and calls
+`newChannelBuilder()`, mirroring its current behavior.
+
+The new ChannelCredentials API cannot be mixed with the pre-existing
+`useTransportSecurity()` and `usePlaintext()` builder APIs. The old-style
+methods will throw if called on a builder with a ChannelCredential provided.
+
+Note that this new API does not allow: 1) creating a channel without a
+credential and supplying the credential later and 2) changing the credential
+after creating the builder. This is not expected to be much of an issue as
+users typically have the security information available before creating the
+builder and reusing a builder is rare, especially to change the security. 
+
+```java
+package io.grpc;
+
+/**
+ * Represents a security configuration to be used for channels. There is no
+ * generic mechanism for processing arbitrary ChannelCredentials; the consumer
+ * of the credential (the channel) must support each implementation explicitly
+ * and separately. Consumers are not required to support all types or even all
+ * possible configurations for types that are partially supported, but they
+ * <em>must</em> at least fully support ChoiceChannelCredentials.
+ *
+ * <p>A Credential provides client identity and authenticates the server. This
+ * is different from CallCredentials, which only provides client identity. They
+ * can also influence types of encryption used and similar security
+ * configuration.
+ */
+public abstract class ChannelCredentials {}
+
+public final class TlsChannelCredentials extends ChannelCredentials {
+  /* Complicated. Described separately below. */
+}
+
+/** No client identity, authentication, or encryption is to be used. */
+public final class InsecureChannelCredentials extends ChannelCredentials {}
+
+/**
+  * Provides a list of ChoiceChannelCredentials, where any one may be used. The
+  * credentials are in preference order.
+  */
+public final class ChoiceChannelCredentials extends ChannelCredentials {
+  public ChoiceChannelCredentials(ChannelCredentials... creds) {...}
+
+  public List<ChannelCredentials> getCredentialsList() { return creds; }
```

Comment:
What will happen if no credential ends up being applicable?

Reply:
Then the transport doesn't understand the credential. It would go into the normal flow of allowing other transports a try and if they all fail producing an exception.